### PR TITLE
Manual Network Try 3.

### PR DESCRIPTION
**Pull request by lipere123**
A rework of the manual network.
I hope it will be a bit better.

### CODE REVIEW

**Comment by lipere123** [comment]
@AlexCheema this new manual discovery is bit better I think.
But I have trouble with GRPC that is not connected.
When I check the port, they are open.
Thanks in avance.
Best REgards.
Benjamin.

[edgenode2-exo-run.log](https://github.com/user-attachments/files/17355224/edgenode2-exo-run.log)
[edgenode3-exo-run.log](https://github.com/user-attachments/files/17355223/edgenode3-exo-run.log)

File: exo/main.py [hunk 1]
```diff
@@ -59,9 +62,21 @@
 inference_engine = get_inference_engine(inference_engine_name, shard_downloader)
 print(f"Using inference engine: {inference_engine.__class__.__name__} with shard downloader: {shard_downloader.__class__.__name__}")
 
-if args.node_port is None:
-  args.node_port = find_available_port(args.node_host)
-  if DEBUG >= 1: print(f"Using available port: {args.node_port}")
+if args.discovery_module == "manual":
+  if args.discovery_config is None:
+    args.discovery_module == "udp"
```
Comment (AlexCheema) [inline exo/main.py:67]: What's going on here? Can you please throw an error if `args.discovery_module == "manual"` and `args.discovery_config is None`

File: exo/networking/manual/read_config.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+from exo.topology.device_capabilities import DeviceCapabilities, DeviceFlops
+import socket
+import yaml
+
+class ReadManualConfig():
```
Comment (AlexCheema) [inline exo/networking/manual/read_config.py:5]: Why is this a class? Just make a single function that reads the device capabilities from the file

File: exo/main.py [hunk 3]
```diff
@@ -59,9 +62,21 @@
...
+  else:
+    list_device = ReadManualConfig(discovery_config=args.discovery_config)
```
Comment (AlexCheema) [inline exo/main.py:69]: `ReadManualConfig` should be a function not a class

File: exo/main.py [hunk 4]
```diff
@@ -59,9 +62,21 @@
...
+    list_device.device_capabilities((str((list_device.whoami))))
```
Comment (AlexCheema) [inline exo/main.py:70]: im confused why there's another function call here.

File: exo/main.py [hunk 5]
```diff
@@ -59,9 +62,21 @@
...
+    args.node_id = list_device.node_id
+    args.node_host = list_device.node_host
+    args.node_port = list_device.node_port
+    args.wait_for_peers = 1
+
+if args.discovery_module != "manual":
```
Comment (AlexCheema) [inline exo/main.py:76]: This if shouldn't be here.

File: exo/networking/manual/manual_discovery.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  async def task_discover_peers(self):
+    while True:
+      try:
+        print("task_discover_peers")
+        current_time = time.time()
+
+        for device in self.list_device.config_devices:
+          if f"{self.list_device.whoami}" != f"{device['server']}":
+            print(f"Getting Id {device['id']} == Adresse: {device['address']} {device['port']}")
+            peer_id = (str(f"{device['id']}"))
```
Comment (AlexCheema) [inline exo/networking/manual/manual_discovery.py:64]: what's with the `str(f"` stuff? just `peer_id = device['id']` is fine...

File: exo/networking/manual/manual_discovery.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+  async def task_discover_peers(self):
+    while True:
+      try:
+        print("task_discover_peers")
+        current_time = time.time()
+
+        for device in self.list_device.config_devices:
+          if f"{self.list_device.whoami}" != f"{device['server']}":
+            print(f"Getting Id {device['id']} == Adresse: {device['address']} {device['port']}")
```
Comment (AlexCheema) [inline exo/networking/manual/manual_discovery.py:63]: Please put `if DEBUG >= 2:` for all these prints

File: exo/networking/manual/manual_discovery.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    self.node_id = node_id
+    self.node_port = node_port
+    self.discovery_config = discovery_config
+    self.list_device = ReadManualConfig(discovery_config=self.discovery_config)
+    self.create_peer_handle = create_peer_handle
+    self.discovery_interval = discovery_interval
+    self.discovery_timeout = discovery_timeout
+    self.update_interval = update_interval
+    self.device_capabilities = device_capabilities
+    self.known_peers: Dict[str, Tuple[PeerHandle, float, float]] = {}
```
Comment (AlexCheema) [inline exo/networking/manual/manual_discovery.py:33]: Shouldn't this just be `Dict[str, PeerHandle]` now? I don't think we need the times any more.

File: exo/main.py (hunk 1 above)
Comment (lipere123) [reply exo/main.py]: Done.

File: exo/networking/manual/read_config.py (hunk 2 above)
Comment (lipere123) [reply exo/networking/manual/read_config.py:6]: Frankly, I think it is more flexible.
For exemple, for now, I set args.wait_for_peers = 1 in main.py.
But after we finish debug, I want the number of peer to wait to comme from ReadManualConfig.

File: exo/main.py (hunk 3 above)
Comment (lipere123) [reply exo/main.py]: Frankly, I think it is more flexible.
For exemple, for now, I set args.wait_for_peers = 1 in main.py.
But after we finish debug, I want the number of peer to wait to comme from ReadManualConfig.

File: exo/main.py (hunk 4 above)
Comment (lipere123) [reply exo/main.py]: There is 2 call fo ReadManualConfig.
One in the main for the current instance so the main.py is instancied with his own value.
The second one for the discovery of the other devices.

File: exo/networking/manual/manual_discovery.py (hunk 6 above)
Comment (lipere123) [reply exo/networking/manual/manual_discovery.py]: Done.

File: exo/networking/manual/manual_discovery.py (hunk 7 above)
Comment (lipere123) [reply exo/networking/manual/manual_discovery.py]: Done.

File: exo/networking/manual/manual_discovery.py (hunk 8 above)
Comment (lipere123) [reply exo/networking/manual/manual_discovery.py:34]: I don't know about that, for now I simplified the discovery with one function that input all the other devices on startup.
But I remove disconnection, ... etc for debug purpose.

Frankly it is your call, same for the GRPC problem.
Do you have some discord time to help me on : 
is_connected = await new_peer_handle.is_connected()
health_ok = await new_peer_handle.health_check()
that are not working, please ?

Thanks in advance.
Best Regards.
Benjamin.

File: exo/main.py [hunk 9]
```diff
@@ -59,9 +62,28 @@
...
 print(f"Using inference engine: {inference_engine.__class__.__name__} with shard downloader: {shard_downloader.__class__.__name__}")
 
-if args.node_port is None:
-  args.node_port = find_available_port(args.node_host)
-  if DEBUG >= 1: print(f"Using available port: {args.node_port}")
+try:
+  if args.discovery_module == "manual":
+    if args.discovery_config is None:
+      raise ValueError("--discovery-config is necessary when using --discovery-module manual")
+    else:
```
Comment (AlexCheema) [inline exo/main.py:69]: redundant else

File: exo/main.py [hunk 10]
```diff
@@ -59,9 +62,28 @@
...
+      # Read for the current instance
+      list_device = ReadManualConfig(discovery_config=args.discovery_config)
+      list_device.device_capabilities((str((list_device.whoami))))
+      # Initialisation of the current instance with the first ReadManualConfig, and directly into main.py
+      args.node_id = list_device.node_id
+      args.node_host = list_device.node_host
+      args.node_port = list_device.node_port
+      args.wait_for_peers = 1
+  else:
```
Comment (AlexCheema) [inline exo/main.py:78]: why else? can't the code here also run for manual discovery?

File: exo/networking/manual/manual_discovery.py (hunk 8 above)
Comment (AlexCheema) [reply exo/networking/manual/manual_discovery.py:34]: sure I can help. it's quite simple tho - only healthy peers should be in the topology

File: exo/networking/manual/read_config.py (hunk 2 above)
Comment (AlexCheema) [reply exo/networking/manual/read_config.py:6]: I dont understand what you mean

File: exo/networking/manual/manual_discovery.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        for device in self.list_device.config_devices:
+          if f"{self.list_device.whoami}" != f"{device['server']}":
+            if DEBUG >= 2: print(f"Getting Id {device['id']} == Adresse: {device['address']} {device['port']}")
+            peer_id = device['id']
+            peer_host = device['address']
+            peer_port = device['port']
+            if peer_id not in self.known_peers or self.known_peers[peer_id][0].addr() != f"{peer_host}:{peer_port}":
+              new_peer_handle = self.create_peer_handle(peer_id, f"{peer_host}:{peer_port}", self.list_device.device_capabilities((str((f"{device['server']}")))))
+              try:
+                is_connected = await new_peer_handle.is_connected()
```
Comment (AlexCheema) [inline exo/networking/manual/manual_discovery.py:71]: no need to check is_connected. just health check

**Review by AlexCheema — CHANGES_REQUESTED** [review]

**Comment by AlexCheema** [comment]
Manual networking was already implemented separately.

**Pull request closed without merging**